0010-endpoint-metadata-extension/README.md

```diff
@@ -0,0 +1,687 @@
+---
+name: Endpoint Metadata Extension
+shortname: 10/EPMDX
+status: draft
+editor: Volodymyr Tkhir <vtkhir@cybervisiontech.com>
+---
+
+- [Introduction](#introduction)
+- [Requirements and constraints](#requirements-and-constraints)
+  - [Metadata keys](#metadata-keys)
+  - [Metadata values](#metadata-values)
+  - [Supported operations](#supported-operations)
+- [Design](#design)
+  - [Full metadata update](#full-metadata-update)
+    - [Full metadata update request](#full-metadata-update-request)
+    - [Full metadata update response](#full-metadata-update-response)
+  - [Partial metadata update](#partial-metadata-update)
+    - [Partial metadata update request](#partial-metadata-update-request)
+    - [Partial metadata update response](#partial-metadata-update-response)
+  - [Get metadata](#get-metadata)
+    - [Get metadata request](#get-metadata-request)
+    - [Get metadata response](#get-metadata-response)
+  - [Get metadata keys](#get-metadata-keys)
+    - [Get metadata keys request](#get-metadata-keys-request)
+    - [Get metadata keys response](#get-metadata-keys-response)
+  - [Delete metadata key](#delete-metadata-key)
+    - [Delete metadata key request](#delete-metadata-key-request)
+    - [Delete metadata key response](#delete-metadata-key-response)
+
+
+## Introduction
+
+The Endpoint Metadata Extension (EPMDX) protocol is an endpoint-aware [Kaa Protocol](/0001-kaa-protocol/README.md) extension.
+
+It is intended to manage endpoint metadata.
+
+## Requirements and constraints
+
+### Metadata keys
+EPMDX protocol supports only string type of metadata keys.
+All UTF-8 characters are supported in metadata key names.
+Metadata key must contain at least one non-whitespace character.
```

Comment:
@rasendubi, I think we should not allow the metadata keys without non-whitespace characters at least for the two reasons:

- Display such key-value pairs on UI in endpoint metadata widget will be a real problem. 
- Display such key-value pairs in EPMDX service instance configuration(list of readable and writable keys) on UI will be the same problem. For example, how to distinguish an empty string and string with one space?
- REST API for the EPR service now has a `PUT api/v1/endpoints/{endpointId}/metadata/{metadataKey}` method to update a single endpoint metadata key
If we will allow the empty metadata key, then the request will look like `PUT api/v1/endpoints/{endpointId}/metadata//`. It is really confusing for me.
- EPMDX and EPR primary designed to store a real metadata like `location`, `name`, `device ID`, etc. It makes no sense to build a [trie](https://en.wikipedia.org/wiki/Trie) in the root of metadata.
We will be able to build a trie in the some metadata key:
```
{
"trie": 
  { "": "Root prefix",
    "p": "P prefix"
  },
"otherValue": 1
}
```